allennlp/data/iterators/multiprocess_iterator.py

```diff
@@ -0,0 +1,152 @@
+            # put None on the queue, and then break out of the loop.
+            if instances:
+                make_batches()
+            output_queue.put(None)
+            break
+        else:
+            instances.append(instance)
+            if len(instances) >= max_instances_in_memory:
+                make_batches()
+
+def _queuer(instances: Iterable[Instance],
+            input_queue: Queue,
+            num_workers: int,
+            num_epochs: Optional[int]) -> None:
+    """
+    Reads Instances from the iterable and puts them in the input_queue.
+    """
+    epoch = 0
+
+    while num_epochs is None or epoch < num_epochs:
+        epoch += 1
+        for instance in instances:
+            input_queue.put(instance)
+
+    # Now put a None for each worker, since each needs to receive one
+    # to know that it's done.
+    for _ in range(num_workers):
+        input_queue.put(None)
+
+@DataIterator.register("multiprocess")
+class MultiprocessIterator(DataIterator):
+    """
+    A ``DataIterator`` that uses ``torch.multiprocessing`` to generate tensor dicts
+    using multiple processes. It's currently less full-featured than some of our
+    other data iterators.
+
+    Parameters
+    ----------
+    batch_size : ``int``, optional, (default = 32)
+        The size of each batch of instances yielded when calling the iterator.
+    num_workers : ``int``, optional (default = 1)
+        The number of processes used for generating tensor dicts.
+    max_instances_in_memory : ``int``, optional, (default = 1024)
+        Each tensor-dict-generating process will wait until it has this many
+        instances, then batch and tensorize them. It's probably a good idea
+        to have this be a multiple of ``batch_size``, otherwise you'll end up
+        with a lot of partially-full batches.
+    """
+    def __init__(self,
+                 batch_size: int = 32,
+                 num_workers: int = 1,
+                 max_instances_in_memory: int = 1024) -> None:
+        super().__init__()
+        self.num_workers = num_workers
+        self.batch_size = batch_size
+        self.max_instances_in_memory = max_instances_in_memory
+        self.vocab: Vocabulary = None
+
+        self.processes: List[Process] = []
+        self.queuer: Optional[Process] = None
+
+    def _create_batches(self, instances: Iterable[Instance], shuffle: bool) -> Iterable[Batch]:
+        raise RuntimeError("MultiprocessIterator doesn't use create_batches")
+
+    def __call__(self,
+                 instances: Iterable[Instance],
+                 num_epochs: int = None,
+                 shuffle: bool = True) -> Iterator[TensorDict]:
+
+        # If you run it forever, the multiprocesses won't shut down correctly.
+        # TODO(joelgrus) find a solution for this
+        if num_epochs is None:
+            raise ConfigurationError("Multiprocess Iterator must be run for a fixed number of epochs")
+
+        # TODO(joelgrus) are these the right sizes?
```

Comment:
See comment above -- seems we could also make the input queue size approximately batch_size times larger then output queue size.